// @graphql-toolkit/graphql-file-loader/index.esm.js
import { isValidPath } from '@graphql-toolkit/common';
import { parse, Source } from 'graphql';
import { extname, isAbsolute, resolve } from 'path';
import { existsSync, readFileSync } from 'fs';
import { importSchema } from 'graphql-import';

const GQL_EXTENSIONS = ['.gql', '.graphql', '.graphqls'];
class GraphQLFileLoader {
    loaderId() {
        return 'graphql-file';
    }
    async canLoad(pointer, options) {
        if (isValidPath(pointer)) {
            const extension = extname(pointer).toLowerCase();
            if (GQL_EXTENSIONS.includes(extension)) {
                const normalizedFilePath = isAbsolute(pointer) ? pointer : resolve(options.cwd || process.cwd(), pointer);
                if (existsSync(normalizedFilePath)) {
                    return true;
                }
            }
        }
        return false;
    }
    async load(pointer, options) {
        const normalizedFilePath = isAbsolute(pointer) ? pointer : resolve(options.cwd || process.cwd(), pointer);
        const content = readFileSync(normalizedFilePath, 'utf-8').trim();
        if (content && content !== '') {
            if (!options.skipGraphQLImport && /^\#.*import /i.test(content.trimLeft())) {
                const importedSchema = importSchema(normalizedFilePath);
                return {
                    location: pointer,
                    document: parse(importedSchema),
                };
            }
            else {
                return {
                    location: pointer,
                    document: parse(new Source(content, pointer)),
                };
            }
        }
        return null;
    }
}

export { GraphQLFileLoader };
